### ISSUE TITLE
abel.transform() transform arguments as dict or **kwargs

### ISSUE THREAD

**Issue by DanHickstein** [PyAbel/PyAbel#109]
Currently, the `abel.transform()` function is defined like this:
```python
def transform(
    IM, direction='inverse', method='three_point', center='none',
        verbose=True, symmetry_axis=None,
        use_quadrants=(True, True, True, True),
        transform_options={}, center_options={}):
```

which means that it accepts `dict` arguments for the keyword arguments that will be passed to the transform functions and to the centering functions. Like this for example:

```python
AQ1 = abel_transform[method](Q1, direction=direction, **transform_options)
```
This leads to a somewhat clunky syntax if the user wants to specify arguments to these sub-functions:

```python
abel.transform(data, method='basex', 
             transform_options={ 'basis_dir' : '/Users/mydir', 'dr ': 0.1 } 
             center_options={ 'center' : (250, 264) } )
```

In https://github.com/PyAbel/PyAbel/pull/108#issuecomment-188706374, @stggh suggests that we switch to `**kwargs` instead. I think that this would look like:
```python
def transform(
    IM, direction='inverse', method='three_point', center='none',
        verbose=True, symmetry_axis=None,
        use_quadrants=(True, True, True, True),
        **kwargs):
```
where the transform functions are called with: 
```python
AQ1 = abel_transform[method](Q1, direction=direction, **kwargs)
```
The problem is that the `kwargs` for the centering function and the transform functions are lumped together into a big mass of `kwargs` and when we pass `**kwargs` to a transform function, we are also passing additional argument. Thus, each transform function would need to be able to accept any number of keyword arguments. So, we would need to define 

```python
def basex_transform(data, nbf='auto', basis_dir='./', dr=1.0, verbose=True,
                    direction='inverse', **kwargs):
```
where I have added **kwargs, which serves no purpose except to collect excessive keyword arguments that get passed.

The **kwargs method has the following advantages:

1) Allows users to specify the arguments in a simple `<parameter> = <value>` syntax. 

But, it has the following disadvantages compared to the `dict` method:

1) If a user supplies a keyword that is not relevant to a transform method, they should get an error. This occurs with the `dict` method, but does not occur with the **kwargs method. 

2) `**kwargs` method requires that we include `**kwargs` in all functions called by the `abel.transform()` method. This is not really a problem, but annoying and could mask errors. 

3) The `dict` method makes it explicit which arguments are being passed to which function, reducing confusion and bugs.

One compromise would be to keep the dictionaries for passing arguments to the centering and transform functions, but also to make the most useful kwargs actual keyword arguments of the `abel.transform()` function. For example, we are already doing this with the `center='com'` keyword. This is then passes to the centering function. We could do the same thing with additional keyword arguments of the transform functions that we thought were especially useful. For example, I think that all of the transform methods accept the `dr=1` parameter, so perhaps we should also make this a keyword argument for the `abel.transform()` function.

TLDR: I think that the code will be significantly cleaner if we stick with passing `dict`s. The only disadvantage is that the learning curve is a little bit steeper for users. If there is a commonly used keyword argument for the transform or centering functions, it's no problem to just add it as a keyword argument for the `abel.transform()` function.

**Comment by ldes89150** [comment]
I like the `**kwargs` syntax.

Perhaps we can check it with something like:

```python
for vk in validkwargs:
    if vk not in kwargs:
        raise ValueError()
```

**Comment by DanHickstein** [comment]
Yes, there are ways to filter through the **kwargs. They get a little confusing. Note that the first comment on this SO post is that a dictionary should probably be used instead: http://stackoverflow.com/questions/23430248/separating-kwargs-for-different-functions

**Comment by DhrubajyotiDas** [comment]
>TLDR: I think that the code will be significantly cleaner if we stick with passing `dicts`.

This makes me lean towards the dict approach. I think we can make it work with good examples/sensible defaults.

**Comment by DanHickstein** [comment]
Actually, the same technique is used to set the `arrowprops` in matplotlib's `annotate` function: 
http://matplotlib.org/users/annotations_intro.html

From this, I learned that we can actually make the `dict` syntax prettier:
```python
abel.transform(data, method='basex', 
             transform_options=dict(basis_dir= '/Users/mydir', dr=0.1),
             center_options=dict(center =(250, 264)) )
```

**Comment by ldes89150** [comment]
> ```python`
>abel.transform(data, method='basex', 
>            transform_options = dict(basis_dir= '/Users/mydir', dr=0.1),
>           center_options    = dict(center=(250, 264)) )
>```
looks nice, too

**Comment by DanHickstein** [comment]
...and I guess that we could just add `crop` as a keyword of `abel.transform`, and then nobody would really need to use the `center_options

**Comment by DhrubajyotiDas** [comment]
>I guess that we could just add `crop` as a keyword of `abel.transform`

**Comment by DanHickstein** [comment]
I realized that `center` was already a keyword argument of `abel.transform`, so the above example that I was using was a bit silly. If we have both `center` (which accepts a tuple for a "bespoke" center as well as `str` arguments to use automatic centering) and `crop` as keywords for `abel.transform()`, then there are not many other options that could potentially be passes to the `center_image` function. Potentially there are, and we can keep the `center_options={}` keyword, but if we have `center` and `crop`, then that should take care of 99% of all scenarios.

**Comment by DhrubajyotiDas** [comment]
Ah, that makes a lot of sense. Yes, with `center` and `crop` as keywords, most cases are indeed covered.

**Comment by stggh** [comment]
I am extremely impressed with this syntax
>abel.transform(data, method='basex', 
             transform_options = dict(basis_dir= '/Users/mydir', dr=0.1),
             center_options    = dict(center=(250, 264)) )

and it appears to work. I am able to pass `dict(dr=0.5)` to `abel.transform()`.

Re the centering functions in `abel.tools.center.py` and #103 point 4.
>Yes, my thought was to have several find_center... functions that only find the center (and don't return the centered image). Actually setting the center of the image brings up the issue of cropping the image, and this should be treated consistently but the new set_center function that @ldes89150 has expanded in his PR. In summary:
tools.center.find_center() function that takes an image and calls any one of the
find_center_... functions. The center of the image can then be set with the
set_center() function.
center_image() combines the find_center and set_center functions, and is probably what most users will use.

Q: @ldes89150 (your baby), @DanHickstein, @DhrubajyotiDas (for ideas) at the moment is there a function that will return an odd size image from an even image input, recentered, without using the centering (optimization) functions?

This question comes about because I am attempting to ensure that `symmetry.find_image_center_by_slice()` returns only the image center coordinates/origin. However, I am not sure what function to call next to return an odd-centered image.

**Comment by DanHickstein** [comment]
I think the idea was that you could call `set_center(IM, center=(100,100))` and it would return an odd image. Is that not the case?

**Comment by stggh** [comment]
Not from what I (learner driver) can tell:
```python
In [13]: y = np.ones((4,4))

In [14]: y
Out[14]: 
array([[ 1.,  1.,  1.,  1.],
       [ 1.,  1.,  1.,  1.],
       [ 1.,  1.,  1.,  1.],
       [ 1.,  1.,  1.,  1.]])

In [22]: abel.tools.center.set_center(y, center=(1.5, 1.5))
delta = (0.5, 0.5)
Out[22]: 
array([[ 0.,  0.,  0.,  0.],
       [ 0.,  1.,  1.,  1.],
       [ 0.,  1.,  1.,  1.],
       [ 0.,  1.,  1.,  1.]])

In [23]: abel.tools.center.set_center(y, center=(1.5, 1.5), crop='valid_region')
delta = (0.5, 0.5)
Out[23]: 
array([[ 1.,  1.],
       [ 1.,  1.]])

```

**Comment by ldes89150** [comment]
I thought it is implemented in `get_image_quadrants` and `put_image_quadrants` because the `iabel_` function requires the symmetry axis at first column. The `get_image_quadrants` can generate suitable quadrant for the transform function

**Comment by ldes89150** [comment]
If `set_center` is required to return an `odd` image, the `crop = "maintain_size"` would be nonsense.

**Comment by stggh** [comment]
No worries. PyAbel requires an easy method for the user to convert their even size (width) image to an odd size. The current Abel transforms (except onion_peeling) require the left-most pixel to be the image center-line. What is the simplest method to achieve this?

I have been using:   `IMx = scipy.nidimage.interpolation.shift(IM, (-0.5, -0.5))[:-1, :-1]`
Are we able to implement the same or a better function in `center.py`? 
In principle, the image need only be centered in width.

`get_image_quadrants` only splits the image. Re-centering requires user choice of method, so this functionality may be better handled in `center.py`.

**Comment by DanHickstein** [comment]
The maintain-size keyword would still make some sense, even if odd images
are returned. The function could still approximately maintain the size and
just return the closest odd sized image. I suggest that we add an
odd_size=True keyword to set_center.

On Thursday, February 25, 2016, Steve Gibson <notifications@github.com>
wrote:

> No worries. PyAbel requires an easy method for the user to convert their
> even size (width) image to an odd size. The current Abel transforms (except
> onion_peeling) require the left-most pixel to be the image center-line.
> What is the simplest method to achieve this?
>
> I have been using: IMx = scipy.nidimage.interpolation.shift(IM, (-0.5,
> -0.5))[:-1, :-1]
> Are we able to implement the same or a better function in center.py?
> In principle, the image need only be centered in width.
>
> get_image_quadrants only splits the image. Re-centering requires user
> choice of method, so this functionality may be better handled in center.py
> .
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/PyAbel/PyAbel/issues/109#issuecomment-189034505>.
>


-- 
Daniel D. Hickstein
DanHickstein@gmail.com
Phone: 301 787 1853

**Comment by DanHickstein** [comment]
It looks like this has been resolved. Closing this issue.
